br/pkg/streamhelper/advancer.go

```diff
@@ -0,0 +1,428 @@
+	defer c.recordTimeCost("try advance", zap.Uint64("checkpoint", rst.TS), zap.Int("len", len(rst.Ranges)))()
+	ranges := CollapseRanges(len(rst.Ranges), func(i int) kv.KeyRange { return rst.Ranges[i] })
+	workers := utils.NewWorkerPool(4, "subranges")
+	eg, cx := errgroup.WithContext(ctx)
+	collector := NewClusterCollector(ctx, c.env)
+	collector.setOnSuccessHook(c.cache.InsertRange)
+	for _, r := range ranges {
+		r := r
+		workers.ApplyOnErrorGroup(eg, func() error {
+			defer c.recordTimeCost("get regions in range", zap.Uint64("checkpoint", rst.TS))()
+			return c.GetCheckpointInRange(cx, r.StartKey, r.EndKey, collector)
+		})
+	}
+	err := eg.Wait()
+	if err != nil {
+		c.cache.InsertRanges(*rst)
+		return err
+	}
+
+	result, err := collector.Finish(ctx)
+	if err != nil {
+		c.cache.InsertRanges(*rst)
+		return err
+	}
+	fr := result.FailureSubranges
+	if len(fr) != 0 {
+		log.Info("failure regions collected", zap.Int("size", len(fr)))
+		c.cache.InsertRanges(RangesSharesTS{
+			TS:     rst.TS,
+			Ranges: fr,
+		})
+	}
+	return nil
+}
+
+func (c *CheckpointAdvancer) UpdateGlobalCheckpointLight(ctx context.Context) (uint64, error) {
+	log.Debug("current tree", zap.Stringer("ct", c.cache))
+	rsts := c.cache.PopRangesWithGapGT(config.DefaultTryAdvanceThreshold)
+	if len(rsts) == 0 {
+		return 0, nil
+	}
+	workers := utils.NewWorkerPool(config.DefaultMaxConcurrencyAdvance, "regions")
+	eg, cx := errgroup.WithContext(ctx)
+	for _, rst := range rsts {
+		rst := rst
+		workers.ApplyOnErrorGroup(eg, func() error { return c.tryAdvance(cx, rst) })
+	}
+	err := eg.Wait()
+	if err != nil {
+		return 0, err
+	}
+	log.Debug("new tree", zap.Stringer("cache", c.cache))
+	ts := c.cache.CheckpointTS()
+	return ts, nil
+}
+
+func (c *CheckpointAdvancer) CalculateGlobalCheckpoint(ctx context.Context) (uint64, error) {
+	var (
+		cp = uint64(math.MaxInt64)
+		// TODO: Use The task range here.
+		thisRun []kv.KeyRange = []kv.KeyRange{{}}
+		nextRun []kv.KeyRange
+	)
+	defer c.recordTimeCost("record all")
+	cx, cancel := context.WithTimeout(ctx, c.cfg.MaxBackoffTime)
+	defer cancel()
+	for {
+		coll := NewClusterCollector(ctx, c.env)
+		coll.setOnSuccessHook(c.cache.InsertRange)
+		for _, u := range thisRun {
+			err := c.GetCheckpointInRange(cx, u.StartKey, u.EndKey, coll)
+			if err != nil {
+				return 0, err
+			}
+		}
+		result, err := coll.Finish(ctx)
+		if err != nil {
+			return 0, err
+		}
+		log.Debug("full: a run finished", zap.Any("checkpoint", result))
+
+		nextRun = append(nextRun, result.FailureSubranges...)
+		if cp > result.Checkpoint {
+			cp = result.Checkpoint
+		}
+		if len(nextRun) == 0 {
+			return cp, nil
+		}
+		thisRun = nextRun
+		nextRun = nil
+		log.Debug("backoffing with subranges", zap.Int("subranges", len(thisRun)))
+		time.Sleep(c.cfg.BackoffTime)
+	}
+}
+
+func CollapseRanges(length int, getRange func(int) kv.KeyRange) []kv.KeyRange {
+	frs := make([]kv.KeyRange, 0, length)
+	for i := 0; i < length; i++ {
```

Comment:
Why not set `ranges []kv.KeyRange` as input arguements directly, but a closure ?

Reply:
At the very beginning, this function may accept `[]Region`, which contains a key range but cannot be passed directly. This signature can handle that condition. And for now I think keep this is still OK.